llvm/lib/Target/AMDGPU/AMDGPUInsertDelayAlu.cpp

```diff
@@ -50,8 +50,12 @@ class AMDGPUInsertDelayAlu {
   static bool instructionWaitsForSGPRWrites(const MachineInstr &MI) {
     // These instruction types wait for VA_SDST==0 before issuing.
     const uint64_t VA_SDST_0 = SIInstrFlags::SALU | SIInstrFlags::SMRD;
-
-    return MI.getDesc().TSFlags & VA_SDST_0;
+    if (MI.getDesc().TSFlags & VA_SDST_0) {
+      for (auto &Op : MI.uses())
```

Comment:
SALU wait for VA_SDST if they read **or write** a (non-vcc) SGPR, which makes sense to avoid WAW dependencies, so maybe check all `operands()` here.